npm/src/controller/oauth.ts

```diff
@@ -410,12 +443,30 @@ export class OAuthController implements IOAuthController {
       audience: this.opts.samlAudience!,
     };
 
+    if (session && session.redirect_uri && !allowed.redirect(session.redirect_uri, samlConfig.redirectUrl)) {
+      throw new JacksonError('Redirect URL is not allowed.', 403);
+    }
+
     if (session && session.id) {
       validateOpts.inResponseTo = session.id;
     }
 
-    const profile = await validateResponse(rawResponse, validateOpts);
-
+    let profile;
+    try {
+      profile = await validateResponse(rawResponse, validateOpts);
+    } catch (err: unknown) {
+      if (!isIdPFlow && session?.redirect_uri) {
```

Comment:
We should standardize this for both flows by using `(session && session.redirect_uri) || samlConfig.defaultRedirectUrl` to figure out the redirect_uri to take the user to